chrome/content/zotero/xpcom/itemTreeManager.js

```diff
@@ -0,0 +1,267 @@
+			return matches.reduce((acc, match) => acc.concat(match), []);
+		}
+		return allColumns.filter(col => Object.keys(optionOrOptions).every(key => col[key] === optionOrOptions[key]));
+	}
+
+	/**
+	 * A centralized data source for custom columns. This is used by the ItemTreeRow to get data.
+	 * @param {Zotero.Item} item 
+	 * @param {string} field 
+	 * @param {boolean} unformatted 
+	 * @returns 
+	 */
+	getCustomCellData(item, field, unformatted) {
+		if (this._customDataProvider[field]) {
+			return this._customDataProvider[field](item, field, unformatted);
+		}
+		return "";
+	}
+
+	/**
+	 * Get columns by type. Only support itemtree and custom for now.
+	 * @param {"itemtree" | "custom"} type 
+	 * @returns 
+	 */
+	_getColumnsByType(type) {
+		type = type || "itemtree";
+		if (type === "itemtree") {
+			return [...ITEMTREE_COLUMNS];
+		}
+		if (type === "custom") {
+			return [...this._customColumns];
+		}
+	}
+
+	/**
+	 * Check if column option is valid. An atomic operation.
+	 * @param {ItemTreeColumnOption | ItemTreeColumnOption[]} optionOrOptions 
+	 * @returns {boolean}
+	 */
+	_validateColumnOption(optionOrOptions) {
+		/**
+		 * Validate column option. An atomic operation.
+		 * @param {ItemTreeColumnOption | ItemTreeColumnOption[]} optionOrOptions 
+		 * @param {(option: ItemTreeColumnOption) => boolean} validator - A function that returns true if the option is valid
+		 * @returns {boolean}
+		 */
+		function validate(optionOrOptions, validator) {
+			if (Array.isArray(optionOrOptions)) {
+				return optionOrOptions.every(opt => validator(opt));
+			}
+			return validator(optionOrOptions);
+		}
+
+		// Check if the input option has duplicate dataKeys
+		const noInputDuplicates = !Array.isArray(optionOrOptions) || !optionOrOptions.find((opt, i, arr) => arr.findIndex(o => o.dataKey === opt.dataKey) !== i);
+		if (!noInputDuplicates) {
+			Zotero.warn(`ItemTree Column options have duplicate dataKey.`);
+		}
+		const requiredProperties = validate(optionOrOptions, (option) => {
+			const valid = option.dataKey && option.label;
+			if (!valid) {
+				Zotero.warn(`ItemTree Column option ${JSON.stringify(option)} must have dataKey and label.`);
+			}
+			return valid;
+		});
+		const noRegisteredDuplicates = validate(optionOrOptions, (option) => {
+			const valid = this.getColumns({ dataKey: option.dataKey }).length === 0;
+			if (!valid) {
+				Zotero.warn(`ItemTree Column option ${JSON.stringify(option)} with dataKey ${option.dataKey} already exists.`);
+			}
+			return valid;
+		});
+		return noInputDuplicates && requiredProperties && noRegisteredDuplicates;
+	}
+
+
+	/**
+	 * Add a new column or new columns. All options must be valid. An atomic operation.
+	 * @param {ItemTreeColumnOption | ItemTreeColumnOption[]} optionOrOptions - An option or array of options to add
+	 * @returns {boolean} - True if column(s) were added, false if not
+	 */
+	_addColumn(optionOrOptions) {
```

Comment:
This and other functions that allow to add/remove multiple should be in plural. E.g. "_addColumns".